File: 3.Software Technologies/00.SOFTWARE TECHNOLOGIES COURSE 2017/24.ExamPreparationThree/Skeletons/PHP Skeleton/src/AppBundle/Controller/FilmController.php
<?php

namespace AppBundle\Controller;

use AppBundle\Entity\Film;
use Doctrine\DBAL\Types\TextType;
use AppBundle\Form\FilmType;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\Request;

class FilmController extends Controller
{
    /**
     * @param Request $request
     * @Route("/", name="index")
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function index(Request $request)
    {
        // Vzimame si repositorito
        $repo = $this->getDoctrine()->getRepository(Film::class);

        //Vzimame si vsichki filmi
        $films = $repo->findAll();

        //podavame gi na viewto
        return $this->render('film/index.html.twig',['films' => $films]);
    }

    /**
     * @param Request $request
     * @Route("/create", name="create")
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function create(Request $request)
    {
        //1. Pravim is nov film
        $film = new Film();

        //2. Pravim si forma ot tip FilmType klas, i zapisvame vsichko v $film
        $form = $this->createForm(FilmType::class, $film);

        //3. Handvalme si requesta ZADULJITELNO
        $form->handleRequest($request);

        //4. Proverqvame dali formata e validna i submitnata
        if($form->isSubmitted() && $form->isValid())
        {
            //5. Ako formata e validna, proverqvame dali ima prazni poleta
            if($film->getName() == "" || $film->getGenre() == ""
                || $film->getDirector() == "" || $film->getYear() == null)
            {
                //5.1 Ako ima prazno pole ostavqme na sushtta stranica
                return $this->redirect("/create");
            }

            //6. Ako vsichko e ok vzimame si bazata, zapazvame filma i seivame
            $em = $this->getDoctrine()->getManager();
            $em->persist($film);
            $em->flush();

            //7. nakraq se vrushtme na "/"
            return $this->redirect("/");
        }

        //4.1 Ako formata ne e validna se vrushtame na sushtata stranica
        return $this->render("film/create.html.twig", ['form' => $form->createView()] );
	}

    /**
     * @Route("/edit/{id}", name="edit")
     *
     * @param $id
     * @param Request $request
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function edit($id, Request $request)
    {
        //Vzimame filma ot bazata po id
        $film = $this
            ->getDoctrine()
            ->getRepository(Film::class)
            ->find($id);

        //proverqvame dali e null
        if($film == null)
        {
            return $this->redirect("/");
        }

        //Suzdavame forma kato kazvame che shte zapisvame vsichko v $film
        $form = $this->createForm(FilmType::class, $film);

        //Hendelvame requesta ZADULJITELNO
        $form->handleRequest($request);

        //proverqvame formata dali e validna
        if($form->isSubmitted() && $form->isValid())
        {
            //proverqvame dali ima prazni poleta vus formata
            if($film->getName() == "" || $film->getGenre() == ""
                || $film->getDirector() == "" || $film->getYear() == null)
            {
                //Ako ima greshka se vrushtame na sushtoto view ZAPAZVAIKI POLETATA !
                return $this->render('film/edit.html.twig',
                    [
                        'film' => $film,
                        'form' => $form->createView()
                    ]);
            }

            //Ako ne e ""
            $em = $this->getDoctrine()->getManager();
            //POLZVAME MERGE, T.E. SMESVAME TOVA V BAZATA S TOVA V PAMETTA !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            $em->merge($film);
            $em->flush();

            //Redirektvame kum glavnata stranica
            return $this->redirect("/");
        }

        //Ako formata ne e validna se vrushtame na sushtoto view
        return $this->render('film/edit.html.twig',
            [
            'film' => $film,
            'form' => $form->createView()
            ]);

    }

    /**
     * @Route("/delete/{id}", name="delete")
     *
     * @param $id
     * @param Request $request
     * @return \Symfony\Component\HttpFoundation\Response
     */
    public function delete($id, Request $request)
    {
        //1. vzimame taska no purvo ni trqbva repositorito
        $filmRepo = $this->getDoctrine()->getRepository(Film::class);
        $film = $filmRepo->find($id);

        //AKO FILMA E NULL REDIREKTVAME KUM INDEX PAGE-a
        if($film == null)
        {
            return $this->redirectToRoute('index');
        }

        $form = $this->createForm(FilmType::class, $film);
        $form->handleRequest($request);

        if($form->isSubmitted() && $form->isValid())
        {

            //Vzimame bazata
            $em = $this->getDoctrine()->getManager();
            //Triem ot neq taska
            $em->remove($film);
            //Izpulnqvame i zapazvame promenite
            $em->flush();

            //Vrushtame se kum index page-a
            return $this->redirectToRoute('index');
        }

        //2.vrushtame viewto kato mu podavame taska
        return $this->render('film/delete.html.twig',
            [
                'film' => $film,
                'form' => $form->createView()
            ]);

    }
}

File: 3.Software Technologies/00.SOFTWARE TECHNOLOGIES COURSE 2017/24.ExamPreparationThree/Skeletons/Java Skeleton/src/main/java/imdb/controller/FilmController.java
package imdb.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import imdb.bindingModel.FilmBindingModel;
import imdb.entity.Film;
import imdb.repository.FilmRepository;

import java.util.List;

@Controller
public class FilmController {

	private final FilmRepository filmRepository;

	@Autowired
	public FilmController(FilmRepository filmRepository) {

		// VAJNO !!! TUK PROSTO SETVAME FILMREPOSITORITO :
		this.filmRepository = filmRepository;
	}

	@GetMapping("/")
	public String index(Model model) {

		// Vzimame vsichki filmi
		List<Film> films = filmRepository.findAll();

		// Podavame gi na viwto
		model.addAttribute("films", films);

		// Podavame samoto view
		model.addAttribute("view", "film/index");
		return "base-layout";
	}

	@GetMapping("/create")
	public String create(Model model) {

		model.addAttribute("view", "film/create");
		return "base-layout";
	}

	@PostMapping("/create")
	public String createProcess(Model model, FilmBindingModel filmBindingModel) {

		//1.Proverqvame dali ima prazni poleta
		if(filmBindingModel.getName().equals("") || filmBindingModel.getGenre().equals("")
				|| filmBindingModel.getDirector().equals(""))
		{
			// redirektvame kum "/create"
			return "redirect:/create";
		}

		//2.Pravim si nov film
		Film film = new Film();

		//3. I go setvame na BindingModea koito poluchihme
		film.setName(filmBindingModel.getName());
		film.setGenre(filmBindingModel.getGenre());
		film.setDirector(filmBindingModel.getDirector());
		film.setYear(filmBindingModel.getYear());

		//4.seivame v bazata i pravim flush
		filmRepository.saveAndFlush(film);

		//5. redirektvame kum "/"
		return "redirect:/";
	}

	@GetMapping("/edit/{id}")
	public String edit(Model model, @PathVariable int id) {

		// TOZI METOD E SUSHTIQ KATO TOZI NA DELETE !!!
		Film film = filmRepository.findOne(id);

		if(film == null)
		{
			return "redirect:/";
		}

		model.addAttribute("film", film);
		model.addAttribute("view", "film/edit");
		return "base-layout";
	}

	@PostMapping("/edit/{id}")
	public String editProcess(Model model, @PathVariable int id, FilmBindingModel filmBindingModel) {

		//1. Vzimame si filma ot bazata
		Film changedFilm = filmRepository.findOne(id);

		if(changedFilm == null)
		{
			return "redirect:/";
		}

		//2. Setvame mu novite parametri ot filmBindingModel-a
		changedFilm.setName(filmBindingModel.getName());
		changedFilm.setGenre(filmBindingModel.getGenre());
		changedFilm.setDirector(filmBindingModel.getDirector());
		changedFilm.setYear(filmBindingModel.getYear());


		//3. Proverqvame dali NQKOE OT TQH e prazno
		if(changedFilm.getName().equals("") || changedFilm.getGenre().equals("")
				|| changedFilm.getDirector().equals("") || changedFilm.getYear() == 0)
		{

			//Ako ima prazno mqsto si ostavqme na sushtata stranica
			model.addAttribute("film", changedFilm);
			model.addAttribute("view", "film/edit");
			return "base-layout";
		}


		//3. AKO VSICHKO E OK prosto pravim flush na bazata BEZ DA SEIVAME NISHTO
		filmRepository.flush();
		// Moje da se napravi i sus filmRepository.saveAndFlush();


		//4. Vrushtame se kum index stranicata !!!!
		return "redirect:/";

	}

	@GetMapping("/delete/{id}")
	public String delete(Model model, @PathVariable int id) {

		Film film = filmRepository.findOne(id);

		if(film == null)
		{
			return "redirect:/";
		}

		model.addAttribute("film", film);
		model.addAttribute("view", "film/delete");
		return "base-layout";
	}

	@PostMapping("/delete/{id}")
	public String deleteProcess(@PathVariable int id) {

		Film film = filmRepository.findOne(id);

		if(film == null)
		{
			return "redirect:/";
		}

		filmRepository.delete(film);
		filmRepository.flush();

		return "redirect:/";
	}
}
